src/compute/src/sink/correction_v2.rs

```diff
@@ -0,0 +1,1109 @@
+        }
+        if self.chunks.iter().any(|c| Rc::strong_count(c) != 1) {
+            return Err(("cursor on shared chunks", self));
+        }
+
+        let mut chain = Chain::default();
+        let mut remaining = Some(self);
+
+        // We might be partway through the first chunk, in which case we can't reuse it but need to
+        // allocate a new one to contain only the updates the cursor can still yield.
+        while let Some(cursor) = remaining.take() {
+            if cursor.chunk_offset == 0 {
+                remaining = Some(cursor);
+                break;
+            }
+            let update = cursor.get();
+            chain.push(update);
+            remaining = cursor.step();
+        }
+
+        if let Some(cursor) = remaining {
+            for chunk in cursor.chunks {
+                let chunk = Rc::into_inner(chunk).expect("checked above");
+                chain.push_chunk(chunk);
+            }
+        }
+
+        Ok(chain)
+    }
+}
+
+impl<D: Data> From<Cursor<D>> for Chain<D> {
+    fn from(cursor: Cursor<D>) -> Self {
+        match cursor.try_unwrap() {
+            Ok(chain) => chain,
+            Err((_, cursor)) => {
+                let mut chain = Chain::default();
+                chain.push_cursor(cursor);
+                chain
+            }
+        }
+    }
+}
+
+/// A non-empty chunk of updates, backed by a columnation region.
+///
+/// All updates in a chunk are sorted by (time, data) and consolidated.
+///
+/// We would like all chunks to have size [`CHUNK_SIZE_BYTES`], to make it easy for the allocator
+/// to re-use chunk allocations. Unfortunately, the current `TimelyStack`/`ChunkedStack` API
+/// doesn't provide a convenient way to pre-size regions, so chunks are currently only fixed-size
+/// in spirit.
+struct Chunk<D: Data> {
+    /// The contained updates.
+    data: StackWrapper<(D, Timestamp, Diff)>,
+    /// Cached value of the current chunk size, for efficient updating of metrics.
+    cached_size: Option<LengthAndCapacity>,
+}
+
+impl<D: Data> Default for Chunk<D> {
+    fn default() -> Self {
+        let capacity = Self::capacity();
+        Self {
+            data: StackWrapper::with_capacity(capacity),
+            cached_size: None,
+        }
+    }
+}
+
+impl<D: Data> fmt::Debug for Chunk<D> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Chunk(<{}>)", self.len())
+    }
+}
+
+impl<D: Data> Chunk<D> {
+    /// Create a new chunk containing a single update.
+    fn from_update<DT: Borrow<D>>(update: (DT, Timestamp, Diff)) -> Self {
+        let capacity = Self::capacity();
+        let mut data = StackWrapper::with_capacity(capacity);
+
+        let (d, t, r) = update;
+        data.copy_destructured(d.borrow(), &t, &r);
+
+        Self {
+            data,
+            cached_size: None,
+        }
+    }
+
+    /// Return the chunk capacity implied by [`CHUNK_SIZE_BYTES`] and the update size.
+    fn capacity() -> usize {
```

Comment:
When using `Array`, this should be `desired_capacity`, and all capacity checks should be against the array's capacity. Lgalloc returns memory with _at least_ the requested capacity.

Reply:
Changed to use the `SizableContainer` methods instead.